proof-of-spacetime.md

```diff
@@ -0,0 +1,197 @@
+        if isFault {
+            // faulty sector, need to post a fault to the chain and try to recover from it
+            return Fatal("Detected late fault")
+        }
+
+        inclusionProofs[n] = inclusionProof
+		sectorsSorted[i] = sectors[challenge.Sector]
+    }
+
+    // Generate the snark
+    snark_proof := GeneratePoStSnark(sectorSize, challenges, sectorsSorted, inclusionProofs)
+
+    return snark_proof
+}
+```
+
+#### Verification
+
+`VerifyPoSt` is the functional counterpart to `GeneratePoSt`. It takes all of `GeneratePoSt`'s output, along with those of `GeneratePost`'s inputs required to identify the claimed proof. All inputs are required because verification requires sufficient context to determine not only that a proof is valid but also that the proof indeed corresponds to what it purports to prove.
+
+```go
+// Verify a PoSt.
+func VerifyPoSt(sectorSize BytesAmount, sectors SectorSet, seed Seed, proof PoStProof, faults FaultSet) bool {
+    challenges := DerivePoStChallenges(seed, faults, sectorSize, len(sectors))
+    sectorsSorted := []
+
+    // Match up commitments with challenges
+    for i := 0; i < len(challenges); i++ {
+        challenge := challenges[i]
+        sectorsSorted[i] = sectors[challenge.Sector]
+    }
+
+    // Verify snark
+    return VerifyPoStSnark(sectorSize, challenges, sectorsSorted)
+}
+```
+
+
+#### Types
+
+```go
+// The random challenge seed, provided by the chain.
+Seed [32]byte
+```
+
+```go
+type Challenge struct {
+    Sector Uint
+    Leaf Uint
+}
+```
+
+#### Challenge Derivation
+
+```go
+// Derive the full set of challenges for PoSt.
+func DerivePoStChallenges(seed Seed, faults FaultSet, sectorSize Uint, sectorCount Uint) [POST_CHALLENGES_COUNT]Challenge {
+    challenges := []
+
+    for n := 0; n < POST_CHALLENGES_COUNT; n++ {
+        attemptedSectors := {SectorID:bool}
+        while challenges[n] == nil {
+            challenge := DerivePoStChallenge(seed, n, faults, attempt, sectorSize, sectorCount)
+
+            // check if we landed in a faulty sector
+            if !faults.Contains(challenge.Sector) {
+                // Valid challenge
+                challenges[n] = challenge
+            }
+
+            // invalid challenge, regenerate
+            attemptedSectors[challenge.Sector] = true
+
+            if len(attemptedSectors) >= sectorCount {
+                Fatal("All sectors are faulty")
+            }
+        }
+    }
+
+    return challenges
+}
+
+// Derive a single challenge for PoSt.
+func DerivePoStChallenge(seed Seed, n Uint, attempt Uint, sectorSize Uint, sectorCount Uint) Challenge {
+    n_bytes := WriteUintToLittleEndian(n)
+    data := concat(seed, n_bytes, WriteUintToLittleEndian(attempt))
+    challenge_bytes := blake2b(data)
+
+    sector_challenge := ReadUintLittleEndian(challenge_bytes[0..8])
+    leaf_challenge := ReadUintLittleEndian(challenge_bytes[8..16])
+
+    return Challenge {
+        Sector: sector_challenge % sectorCount,
```

Comment:
this does not give a uniform distribution, but in practice it should work for this case, worth creating an issue.

If sectorsCount does not divide the 8 bytes, then distribution is not uniform. For example: say that the sector_challenges goes from 0-20, and sectors_count is 17, then sector 0,1,2 are more likely to be queried.

Check how others do this, ping @irenegiacomelli @ben-a-fisch

Reply:
I think when this came up for EC, we decided it was fine to use division. That would mean dividing the challenge range (0..2^64) into `sector_count` sub-ranges and selecting a sector based on which sub-range the challenge falls in. I don't think introduction of `BigNum` math would be problematic here, especially if we already use that method for consensus.

Reply:
That said, we use `mod NODES` to derive challenges within sectors already, so we tolerate the bias there.

I can see the argument that it's okay when the potential challenge set is 'large' (like node count) but if we are revisiting here, we might want to consider everywhere and be consistent.

Reply:
yeah, lower numbers will be `1 / ( 2^len(randomness) / numSectors)` more likely to be challenged.

In practice, that number is effectively zero.